app/controllers/post.js

```diff
@@ -26,7 +99,60 @@ module.exports = {
       }
     })
   },
-  test: function (req, res, next) {
-    res.json({ success: 'ulllu' })
+  upvote: async (req, res, next) => {
+    const { id, userId } = req.params
+    try {
+      const post = await PostModel.findById(id)
+      if (!post) {
+        return res.status(STATUS.NOT_FOUND).json({ error: 'No post found' })
+      }
+      // CHECKS IF THE USER HAS ALREADY UPVOTED THE COMMENT
+      post.votes.upVotes.users.user.filter(user => {
+        if (JSON.stringify(user) === JSON.stringify(userId)) {
+          return res.status(STATUS.BAD_REQUEST).json({ error: 'Bad request' })
+        }
+      })
+      // CHECKS IF THE USER HAS ALREADY DOWNVOTED THE COMMENT
+      post.votes.downVotes.users.user.filter(user => {
+        if (JSON.stringify(user) === JSON.stringify(userId)) {
+          post.votes.downVotes.count = post.votes.downVotes.count - 1
+          post.votes.downVotes.users.user.remove(user)
+        }
+      })
+      post.votes.upVotes.count = post.votes.upVotes.count + 1
+      post.votes.upVotes.users.user.push(userId)
+      await post.save()
+      res.status(STATUS.OK).json({ post: post, message: 'Success' })
+    } catch (error) {
+      HANDLER.handleError(res, error)
+    }
+  },
+  downvote: async (req, res, next) => {
+    const { id, userId } = req.params
+    try {
+      const post = await PostModel.findById(id)
+      if (!post) {
+        return res.status(STATUS.NOT_FOUND).json({ error: 'No post found' })
+      }
+      // CHECKS IF THE USER HAS ALREADY UPVOTED THE COMMENT
+      post.votes.downVotes.users.user.filter(user => {
+        if (JSON.stringify(user) === JSON.stringify(userId)) {
+          return res.status(STATUS.BAD_REQUEST).json({ error: 'Bad request' })
+        }
+      })
+      // CHECKS IF THE USER HAS ALREADY DOWNVOTED THE COMMENT
+      post.votes.upVotes.users.user.filter(user => {
+        if (JSON.stringify(user) === JSON.stringify(userId)) {
+          post.votes.upVotes.count = post.votes.upVotes.count - 1
+          post.votes.upVotes.users.user.remove(user)
+        }
+      })
+      post.votes.downVotes.count = post.votes.downVotes.count + 1
+      post.votes.downVotes.users.user.push(userId)
```

Comment:
why are we not increasing the downVotes count here?

Reply:
The flow goes this way, first we check if the user has already downvoted the comment, if yes we take that as a bad request and it ends here, if not, then we check if the user has already upvoted the comment, if yes we first remove the upvote and then add the downvote count, if not, we directly add the downvote count.